src/google/cloud/ndb/context.py

```diff
@@ -15,21 +15,60 @@
 """Context for currently running tasks and transactions."""
 
 import collections
+import contextlib
+import threading
 
 from google.cloud.ndb import _datastore_api
 from google.cloud.ndb import _eventloop
 from google.cloud.ndb import exceptions
-from google.cloud.ndb import _runstate
 
 
-__all__ = ["AutoBatcher", "Context", "ContextOptions", "TransactionOptions"]
+__all__ = [
+    "AutoBatcher",
+    "Context",
+    "ContextOptions",
+    "get_context",
+    "TransactionOptions",
+]
 
 
 _ContextTuple = collections.namedtuple(
     "_ContextTuple", ["client", "eventloop", "stub", "batches", "transaction"]
 )
 
 
+class _LocalState(threading.local):
```

Comment:
This diff block is the refactored state management. The stack based approach from ``_runstate`` was abandoned because it couldn't gracefully handle the context switching from passing in and out of different coroutines.

Most of the rest of the differences in this PR are spelling changes as a result of this refactor.